From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <heftig@archlinux.org>
Date: Mon, 13 Jan 2025 23:55:38 +0100
Subject: [PATCH] meson: support os_type=arch and detect Arch Linux

And set its defaults:
 - Numeric UID and GID for the polkitd user
 - Statically socket-activated agent helper
---
 meson.build       | 21 +++++++++++++++++----
 meson_options.txt |  2 +-
 2 files changed, 18 insertions(+), 5 deletions(-)

diff --git a/meson.build b/meson.build
index d5d69f4c0d0f..73c573ea3ada 100644
--- a/meson.build
+++ b/meson.build
@@ -301,73 +301,86 @@ if os_type == ''
     ['debian', '/etc/debian_version'],
     ['gentoo', '/etc/gentoo-release'],
     ['pardus', '/etc/pardus-release'],
+    ['arch', '/etc/arch-release'],
     ['lfs', '/etc/lfs-release'],
   ]
 
   foreach os_path: os_paths
     if run_command('test', '-e', os_path[1], check: false).returncode() == 0
       os_type = os_path[0]
       break
     endif
   endforeach
 
   if os_type == ''
     message('Linux distribution autodetection failed, specify the distribution to target using -Dos_type=')
   endif
 endif
 
 # The default UID differs between distributions, set it accordingly if not specified
 polkitd_uid = get_option('polkitd_uid')
 if polkitd_uid == ''
-  polkitd_uid = '-'
+  if os_type == 'arch'
+    polkitd_uid = '102'
+  else
+    polkitd_uid = '-'
+  endif
 endif
 
 # The default GID differs between distributions, set it accordingly if not specified
 polkitd_gid = get_option('polkitd_gid')
 if polkitd_gid == ''
-  polkitd_gid = '-'
+  if os_type == 'arch'
+    polkitd_gid = '102'
+  else
+    polkitd_gid = '-'
+  endif
 endif
 
 # The default privileged group differs between distributions, set it accordingly if not specified
 privileged_group = get_option('privileged_group')
 if privileged_group == ''
   if os_type == 'debian'
     privileged_group = 'sudo'
   else
     privileged_group = 'wheel'
   endif
 endif
 
 # The way the helper is activated differs between distributions, set it accordingly if not specified
 activated_helper = get_option('activated_helper')
 if activated_helper == ''
-  activated_helper = 'false'
+  if os_type == 'arch'
+    activated_helper = 'true'
+  else
+    activated_helper = 'false'
+  endif
 endif
 
 pam_include = get_option('pam_include')
 if pam_include == ''
   if ['suse', 'solaris', 'openembedded'].contains(os_type)
     pam_conf = {
       'PAM_FILE_INCLUDE_AUTH': 'common-auth',
       'PAM_FILE_INCLUDE_ACCOUNT': 'common-account',
       'PAM_FILE_INCLUDE_PASSWORD': 'common-password',
       'PAM_FILE_INCLUDE_SESSION': 'common-session',
     }
   elif os_type.contains('bsd')
     pam_conf = {
       'PAM_FILE_INCLUDE_AUTH': 'system',
       'PAM_FILE_INCLUDE_ACCOUNT': 'system',
       'PAM_FILE_INCLUDE_PASSWORD': 'system',
       'PAM_FILE_INCLUDE_SESSION': 'system',
     }
   elif os_type == 'lfs'
     pam_conf = {
       'PAM_FILE_INCLUDE_AUTH': 'system-auth',
       'PAM_FILE_INCLUDE_ACCOUNT': 'system-account',
       'PAM_FILE_INCLUDE_PASSWORD': 'system-password',
       'PAM_FILE_INCLUDE_SESSION': 'system-session',
     }
-  #if ['redhat', 'gentoo', 'pardus'].contains(os_type)
+  #if ['redhat', 'gentoo', 'pardus', 'arch'].contains(os_type)
   else
     pam_conf = {
       'PAM_FILE_INCLUDE_AUTH': 'system-auth',
diff --git a/meson_options.txt b/meson_options.txt
index 8f9e48207787..3fd070eeb65c 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -9,7 +9,7 @@ option('privileged_group', type: 'string', value: '', description: 'Group to use
 option('activated_helper', type: 'combo', choices: ['false', 'true', ''], value: '', description: 'Enable socket-activated agent helper (default: false)')
 
 option('authfw', type: 'combo', choices: ['pam', 'shadow', 'bsdauth'], value: 'pam', description: 'Authentication framework (pam/shadow)')
-option('os_type', type: 'combo', choices: ['redhat', 'suse', 'gentoo', 'pardus', 'solaris', 'netbsd', 'lfs', 'openembedded', 'debian', ''], value: '', description: 'distribution or OS')
+option('os_type', type: 'combo', choices: ['redhat', 'suse', 'gentoo', 'pardus', 'solaris', 'netbsd', 'lfs', 'openembedded', 'debian', 'arch', ''], value: '', description: 'distribution or OS')
 
 option('pam_include', type: 'string', value: '', description: 'pam file to include')
 option('pam_module_dir', type: 'string', value: '', description: 'directory to install PAM security module')
